src/server.rs

```diff
@@ -144,9 +142,10 @@ pub struct Status {
 }
 
 pub fn json_reply<T: Serialize>(socket: &mut net::socket::TcpSocket, msg: &T) {
-    let mut u: String<U512> = to_string(msg).unwrap();
-    u.push('\n').unwrap();
-    socket.write_str(&u).unwrap();
+    if socket.can_send() {
```

Comment:
This may result in us silently dropping responses, which I don't think is what we want. What's the reason that the socket is not able to send a response in the first place? Is the TX buffer full?

If so, we need to implement a more robust means of flow control here. Silently discarding the reply doesn't seem like the right approach to me.

Reply:
Thanks for the advice! May I know how `can_send()` (or `may_send()`) is somehow violating the current Request/Response system? Because `may_send()` checks whether or not TX has been closed on the server side, and `can_send()` checks, on top of `may_send()`, whether or not the TX buffer is full, I'm confused why replies would still need to be written to the buffer if the client wouldn't be able to receive anything. I added the `can_send()` check after looking into our other smoltcp projects like [Thermostat](https://git.m-labs.hk/M-Labs/thermostat/src/branch/master/src/main.rs#L205), so I'd like to know why `can_send()` isn't needed in the first place.

I'm also confused why `String<U512>` was used instead of `String<U256>`.

On the other hand, if I keep `json_reply()`'s code unchanged, the Request/Response mechanism still works fine.

Reply:
> I'm also confused why String<U512> was used instead of String<U256>.

The template time (`U256` or `U512`) indicates the size of the buffer backing the string (e.g. 256 or 512 bytes). It's an arbitrary selection.

> May I know how can_send() (or may_send()) is somehow violating the current Request/Response system?

Placing a check in `json_reply()` is not the right approach because we are actively trying to respond to a request that we received. `can_send()` may return false simply because the TX buffer is full. With the code as written, we silently ignore this failure, but the client never receives a reply to the request that they sent and may hang forever.

I get that there's no real way around it, but we should at least log a warning here if we can't send the message. Additionally, this is actually a redundant check. We can tell if the socket was able to send the reply by checking the status of `socket.write_str(&u)` (instead of unwrapping it) and use the resulting `Result` to detect a failure to transmit the response.